### DISCUSSION TITLE
How to integrate Eclipse Plugins into EMF model server?

### DISCUSSION THREAD

**Discussion by ZhaolinFang** [eclipse-emfcloud/emfcloud#212] [ModelServer]
Hi,

I am now working on a project which will use the EMF.cloud model server and will add some plugins to the server, I am confused about how to integrate eclipse plugins into the model server like [org.eclipse.emfcloud.modelserver.edit](as I see, this is also a plugin project used by the server). Which files can I modify to integrate my own plugins into the server? Thank you in advance. 

Best regards,
Zhaolin

**Comment by ndoschek** [comment]
Hi @ZhaolinFang,

Thank you for interest in the EMF cloud Model Server :tada:!

Basically, we support consuming the EMF cloud Model Server dependencies either via:
- via maven modules via [the maven central repository](https://central.sonatype.com/search?q=g:org.eclipse.emfcloud.modelserver&smo=true)
- or plugins via [the p2 update sites](https://download.eclipse.org/emfcloud/modelserver/p2/staging/0.8.0.RC05/)

In general, I suggest to have a look at [this EMF cloud project-template](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/tree/main/project-templates/modelserver-glspjava-emf-theia), which sets up a simple application using the Model Server (and the diagram framework GLSP) via maven modules.
If you prefer to consume plugins, you can have a look how we do that for [the example Model Server](https://github.com/eclipse-emfcloud/emfcloud-modelserver/tree/master/examples).

HTH, best regards,
Nina

**Reply by ZhaolinFang** [reply]
Hi @ndoschek,

Thanks for your quick response, I was not able to install P2 update sites in Eclipse either with 0.8.0.RC04 nor 0.8.0.RC05. 
<img width="1356" alt="截屏2023-06-12 21 18 44" src="https://github.com/eclipse-emfcloud/emfcloud/assets/90061850/5c74a326-2407-42f5-aa2e-e54264d47538">.

Besides, I am still confused about my question. Specifically, my project is to build several plugins that run in the model server when the server is launched. How can I connect/combine the plugin I created to the model server? The project structure should look like this:

- EMF Model Server: run MyPlugins when the server is launched, and execute the corresponding function in MyPlugins when needed.
- MyPlugin1: Eclipse Plugin that serves the model server.
- MyPlugin2: Eclipse Plugin that serves the model server.

From my understanding of your previous answer, should I register those MyPlugins somewhere in the model server project or define them somewhere in the model server? Thanks for your time, I really appreciate your help!

Best regards,
Zhaolin

**Reply by ndoschek** [reply]
Hi @ZhaolinFang,

Usually, if you would like to setup your project via p2 update sites, you will need to define a target platform in your workspace which declares the plugins to be pulled. 
E.g. you could have a look at this (unfortunately rather outdated) project: https://github.com/eclipsesource/uml-glsp/tree/master/server
Here we have a diagram editor using GLSP and Model Server via a targetplatform.

If you have a look at the [definition file of the targetplatform](https://github.com/eclipsesource/uml-glsp/blob/master/server/targetplatform/targetplatform.tpd)  you can see which Model Server dependencies will be pulled. (Again, rather outdated, but just to showcase this).
In the next step, your own implementation of the Model Server (that for example defines which resources can be loaded, adds custom commands an so on) defines via the [MANIFEST](https://github.com/eclipsesource/uml-glsp/blob/master/server/com.eclipsesource.uml.modelserver/META-INF/MANIFEST.MF) which plugins should be consumed.
In this case, it consumes the Model Server and UML plugins among others.
Here you could also consume custom plugins contained in the project for example.

Hope I could help to clear things up a bit, but please let me know if there is still anything unclear.
Best, Nina

**Comment by eneufeld** [comment]
Please reopen if issue still persists
